Language plugin: x-default language

andrepereiradasilva opened the pull request:
### Description

This patch adds a new `link` tag in the `head` section with `hreflang` attribute set to `x-default`.
### How to test
1. Fresh Joomla install with multilanguage.
2. Apply https://github.com/joomla/joomla-cms/pull/8524 patch and then this patch.
3. Go to language plugin and choose:
   - Item Associations: Yes
   - Add Alternate Meta Tags: Yes
   - Add x-default Meta Tag: Yes
   - X-default language: Choose "Default frontend option" or any language you want as x-default
4. Create language associations between menu items and content
5. Go to frontend, any page with a language association (in all languages) and check the code (`head` tag). There should be a `link`tag with `hreflang` attribute with `x-default` value and the `href`should point to the same content in the language that you selected in "x-default language" field.
6. Play with the plugin options and check if the x-default tag in there or not according to the options you selected.
### B/C

None that i know of.
### More info

> For language/country selectors or auto-redirecting homepages, you should add an annotation for the hreflang value "x-default" as well:
> <link rel="alternate" href="http://example.com/" hreflang="x-default" />

Source: https://support.google.com/webmasters/answer/189077?hl=en

Commit:
Merge remote-tracking branch 'refs/remotes/joomla/staging' into staging

Commit:
Add x-default hreflang for default site URI

Since Joomla makes a 303 redirect to main language URI, we should add a
x-default herflang in the home page.

See https://support.google.com/webmasters/answer/189077

Commit:
Don't add x-default if default language doesn't have language prefix

infograf768 commented:
Ar reading this: 
http://googlewebmastercentral.blogspot.fr/2013/04/x-default-hreflang-for-international-pages.html

I understand the instructions above as using as a new alternate with x-default a chosen language whether using URL Language code or not.
That language could be the site default language per default, but I guess it should be chosen by the site maintainer, i.e. would need a new parameter in the plugin.

andrepereiradasilva commented:
> I understand the instructions above as using as a new alternate with x-default a chosen language whether using URL Language code or not.

Ok, you think it should be active whether "Remove URL Language Code" is set to "Yes" or "No".
If i do that, when  "Remove URL Language Code" is set to "Yes" it will overwrite the site default language hreflang (ex: en-GB).

> That language could be the site default language per default, but I guess it should be chosen by the site maintainer, i.e. would need a new parameter in the plugin.

I guess you are saying put a parameter to choose the x-default language in language plugin.
I'm no expert in Joomla. But what i understand is:
1. The default language for the 'site' is selected by "star" a language in Extensions -> Language(s) -> Installed - Site. That should be the language served when the user language does not exits, right?

2. If a user point to 'scheme://domain.tld/' Joomla will 303 redirect the user to the default language right? So if x-default language value is 'scheme://domain.tld/' it will be pointing to the default language.

If i'm correct, i don't see the need for another parameter in the plugin. But i could be missing something.

infograf768 commented:
Will try to precise the way I see it:
x-default alternate is a parameter that lets Google define for SEO purposes a certain url as the url displaying stuff specifically aimed at user who would not be specially redirected to one of the possible languages (for example when Language Selection for new Visitors is set to Browser settings on their browsers.)
It is NOT a redirection per se. It is just used to crawl contents. It does not overrides other alternates, it ADDS to the existing alternates and can be the same url as one of the alternates.

Whether this Content Language is the default site language or not, the user should decide.

Example: I have a multilingual site where the default site language is French, I have also English (en-GB) and Spanish.
I may want the English language to be the x-default for Google but I still want the default site language OR the user's browser settings (or cache) to load the correct page.

In this case, when URL Language code is removed, it is the French language which corresponds to the pure `scheme://domain.tld/`.
As I want English to be, for SEO crawling purposes, the x-default alternate, the url should include the url language code (en in my case) `scheme://domain.tld/en/` if SEF is on. 
If I choose French as my x-default, then it will be the pure `scheme://domain.tld/`.

Therefore, we would need a new parameter to choose that language.

[The next question, which is really unclear on Google Webmaster, concerns adding the x-default for Home pages only or all pages. 
Also, I am not totally sure it is necessary in Joomla multilingual case.]

richard67 commented:
As far as I understand, the x-default alternate link should always point to the site root, because this is the only URL from which Joomla! redirects to some language's home page depending on the settings (default or browser language detection), and the link can be added to all pages of the site.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8442">issues.joomla.org/joomla-cms/8442</a>.</sub>

andrepereiradasilva commented:
@infograf768 

> x-default alternate is a parameter that lets Google define for SEO purposes a certain url as the url displaying stuff specifically aimed at user who would not be specially redirected to one of the possible languages (for example when Language Selection for new Visitors is set to Browser settings on their browsers.)

Yes, i know. x-default is only for crawlers for SEO purposes.

> It is NOT a redirection per se. It is just used to crawl contents. It does not overrides other alternates, it ADDS to the existing alternates and can be the same url as one of the alternates.

When i has referring «it will overwrite the site default language hreflang (ex: en-GB)» i has referring to the function `addHeadLink` (https://github.com/joomla/joomla-cms/blob/staging/libraries/joomla/document/html/html.php#L277-L284) that only allows one link tag per URI.

> Whether this Content Language is the default site language or not, the user should decide.
> Example: I have a multilingual site where the default site language is French, I have also English (en-GB) and Spanish.
> I may want the English language to be the x-default for Google but I still want the default site language OR the user's browser settings (or cache) to load the correct page.
> In this case, when URL Language code is removed, it is the French language which corresponds to the pure scheme://domain.tld/.
> As I want English to be, for SEO crawling purposes, the x-default alternate, the url should include the url language code (en in my case) scheme://domain.tld/en/ if SEF is on. 
> If I choose French as my x-default, then it will be the pure scheme://domain.tld/.
> Therefore, we would need a new parameter to choose that language.

Hum, i understand now what you mean.

> [The next question, which is really unclear on Google Webmaster, concerns adding the x-default for Home pages only or all pages. 

I think it should  be in all pages.

@richard67
 
> As far as I understand, the x-default alternate link should always point to the site root, because this is the only URL from which Joomla! redirects to some language's home page depending on the settings (default or browser language detection), and the link can be added to all pages of the site.

I think the hreflang if for the alternate version of the page the crawler is reading.
For instance, if the crawler is reading page "content 1" in en-GB it expects a fr-FR alternative of "content 1" page, a es-ES alternative of "content 1" page, and a fr-FR (in this case x-default) of that page, not of the homepage.

So, in my opinion, this is what is needed for this to work:

- Improvements to `addHeadLink` so it's possible to had several links with same href but different attributes
- In Languague plugin a select box with Use Default languae (pre slected) + the languages 
- Add a link hreflang=x-default in all pages with associations to the href in x-default language

richard67 commented:
@andrepereiradasilva Yes, i just read [http://googlewebmastercentral.blogspot.fr/2013/04/x-default-hreflang-for-international-pages.html](http://googlewebmastercentral.blogspot.fr/2013/04/x-default-hreflang-for-international-pages.html "http://googlewebmastercentral.blogspot.fr/2013/04/x-default-hreflang-for-international-pages.html") again and see it can be done in both ways, like I wrote (see example there, would be a minimal solution) and like you describe it (would be the complete solution but as you mentioned it requires the improvement of addHeadLink and the way how the head links are stored in a list). I was maybe mislead by their example and that they always write "that this page doesn’t target any specific language or locale and is the default page when no other page is better suited", which made me think it should be the home page only, but the more I think it over, the more I favor the complete solution, which requres what you describe ... but the change of the header link handling could be a bigger thing because currently they are stored in an associative array with the URL as key, as far as I remember. I always asked myself who ever could have this crazy idea, because it never was forbidden by any standards to have several head links pointing to the same url for different purpose.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8442">issues.joomla.org/joomla-cms/8442</a>.</sub>

infograf768 commented:
Therefore we have here 3 aspects to consider:
1. Remove URL Language Code should anyway not be a parameter as well as SEF on or not.
2. If each page should have a x-default, what are we going to do if there is no association? Our multilingual implementation allows this.
3. What would happen when we get on a page (drilldown) which does not have a specific menu item?

imho any non-automatic solution would be unmanageable.

andrepereiradasilva commented:
My opinion:

> 1. Remove URL Language Code should anyway not be a parameter as well as SEF on or not.

I'm not quite sure what you mean here.
"Remove URL Language Code should anyway not be a parameter" - Remove the "Remove URL Language Code" in the plugin?
"as SEF on or not" - Remove the "SEF" on or off in global config?
Don't agree with both. I websites with on and others with off, depending on whats needed.

> 2. If each page should have a x-default, what are we going to do if there is no association? Our multilingual implementation allows this.

The same this we do for the other hrelang in pages that don't have association. Don't put it.

> 3. What would happen when we get on a page (drilldown) which does not have a specific menu item?

Articles and categories can be associated to so, like "2."

Another thing, i think in "System - Language Filter" plugin "Add Alternate Meta Tags" should depend on "Item Associations" set to "Yes".

Commit:
Add option in language filter for choose x-default language

For now, since addHeadLink only allows one URL it overwrites the
language hreflang. Changes in addHeadLink  needed

Commit:
ups ... one line missing

Commit:
Also add option to choose current default language

Commit:
cs

andrepereiradasilva commented:
@infograf768 can you check now?
I think the only thing missing is "Improvements to addHeadLink so it's possible to had several links with same the same href but different attributes" but i don't know how to do this without breaking B/C.

Also i notice if i use showon field attribute inside another field already in showon (https://github.com/andrepereiradasilva/joomla-cms/blob/hreflang-x-default/plugins/system/languagefilter/languagefilter.xml#L63) it doesn't work properly.

Commit:
optimization just one line needed

Commit:
Check if is set first ...

Commit:
Final changes for now ... i hope

Commit:
ups

Commit:
cs

Commit:
another cs

infograf768 commented:
Much better. :smile:
Remarks:
Your patch forces a x-default, i.e. there is no Option to not use it although we have
`PLG_SYSTEM_LANGUAGEFILTER_FIELD_XDEFAULT_DESC="This option will add x-default hreflang link tag to improve SEO"`

I suggest to add a new field to set the Option (using this desc) and change the desc for the x-default field  and change code accordingly with a new conditional.

I confirm the issue with showon. Would be nice if we could correct this. 
I ask @dgt41 if he can solve. :smile:

In the meanwhile you could use `showon="alternate_meta:1"` for the 3 fields.

andrepereiradasilva commented:
> Your patch forces a x-default, i.e. there is no Option to not use it although we have
PLG_SYSTEM_LANGUAGEFILTER_FIELD_XDEFAULT_DESC="This option will add x-default hreflang link tag to improve SEO"
> I suggest to add a new field to set the Option (using this desc) and change the desc for the x-default field and change code accordingly with a new conditional.
ok will do later

> I confirm the issue with showon. Would be nice if we could correct this. 
> I ask @dgt41 if he can solve. :smile:
Not yet complete, but already working on a solution (https://github.com/andrepereiradasilva/joomla-cms/commit/5a2086ba52914767eda4873c34934dc7d7ac686e).

andrepereiradasilva committed:
Add new option. thanks @infograf768. Uses new showon mecanism

Test with https://github.com/joomla/joomla-cms/pull/8524

andrepereiradasilva committed:
ups

andrepereiradasilva commented:
Test instructions updated

brianteeman commented on administrator/language/en-GB/en-GB.plg_system_languagefilter.ini:
```diff
@@ -15,6 +15,11 @@ PLG_SYSTEM_LANGUAGEFILTER_FIELD_DETECT_BROWSER_DESC="Choose site default languag
 PLG_SYSTEM_LANGUAGEFILTER_FIELD_ITEM_ASSOCIATIONS_DESC="This option will allow item associations when switching from one language to another."
 PLG_SYSTEM_LANGUAGEFILTER_FIELD_ITEM_ASSOCIATIONS_LABEL="Item Associations"
+PLG_SYSTEM_LANGUAGEFILTER_FIELD_XDEFAULT_DESC="This option will add x-default hreflang link tag to improve SEO."
+PLG_SYSTEM_LANGUAGEFILTER_FIELD_XDEFAULT_LABEL="Add x-default Meta Tag"
```
To comply with the style tag this should be "meta tag" all lowercase - thanks

andrepereiradasilva committed:
lowercase

infograf768 commented:
I have tested this item :white_check_mark: successfully on f83dbd33437af9913c695d6b63c7d1535a7646fa<br /><br />Tested with #8524 and works fine.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8442">issues.joomla.org/joomla-cms/8442</a>.</sub>

anibalsanchez commented:
I have tested this item :white_check_mark: successfully on f83dbd33437af9913c695d6b63c7d1535a7646fa<br /><br />Test OK<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8442">issues.joomla.org/joomla-cms/8442</a>.</sub>

infograf768 commented:
This patch would be for 3.6.0, and AFTER https://github.com/joomla/joomla-cms/pull/8524 is merged.

brianteeman commented:
Thanks - set to RTC - (not sure why it would not be included until 3.6 but I guess thats for a maintainer to decide)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/8442">issues.joomla.org/joomla-cms/8442</a>.</sub>

infograf768 commented:
>not sure why it would not be included until 3.6

because https://github.com/joomla/joomla-cms/pull/8524 has been set to 3.6.0...

wilsonge commented:
I have merged #8524 - can we have one more test with this pr against staging to confirm this still works. and we can merge this

andrepereiradasilva commented:
i tested with latest staging and it work. But let @infograf768 test.

infograf768 commented:
tomorrow i will. i just could see that the "language" type is wrong. The type should be content_language.

andrepereiradasilva commented:
@infograf768 i didn't understand what you mean.

infograf768 commented:
>@infograf768 i didn't understand what you mean.

Oops, in fact you did it.
`name="xdefault_language" type="contentlanguage"`

Just a small cs for the en-GB lang strings values:
Change:
`Add x-default meta tag` to `Add x-default Meta Tag`
and
`X-default language` to `X-default Language`

infograf768 commented:
@wilsonge
After the corrections in the language strings, this is ready to go. :)

andrepereiradasilva commented:
> X-default language to x-default Language

The "X" is for the label title case in this one.

infograf768 commented:
`x-default` is a sort of tag. I do not think we should capitalise it.

@brianteeman 
Could you give your insight on this?

brianteeman commented:
This tag was created by google. I can find no example of where it is ever
capitalised so i would agree it should not be capitalised in the desc

On 18 January 2016 at 17:48, infograf768 <notifications@github.com> wrote:

> x-default is a sort of tag. I do not think we should capitalise it.
>
> @brianteeman <https://github.com/brianteeman>
> Could you give your insight on this?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/8442#issuecomment-172603917>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

andrepereiradasilva committed:
text corrections

andrepereiradasilva commented:
Done.

infograf768 commented:
@wilsonge
This can now be merged. Thanks.
